src/utils/modules.js

```diff
@@ -208,95 +144,85 @@ export class ModuleTracer extends Tracer {
     newTarget: void | ObjectValue,
     performCall: () => Value
   ): void | Value {
-    if (
-      F === this.modules.getRequire() &&
-      !this.modules.disallowDelayingRequiresOverride &&
-      argumentsList.length === 1
-    ) {
+    let requireInfo = this.modules.getRequireInfo();
+    if (requireInfo !== undefined && F === requireInfo.value && argumentsList.length === 1) {
       // Here, we handle calls of the form
       //   require(42)
 
       let moduleId = argumentsList[0];
       let moduleIdValue;
       // Do some sanity checks and request require(...) calls with bad arguments
-      if (moduleId instanceof NumberValue || moduleId instanceof StringValue) {
-        moduleIdValue = moduleId.value;
-      } else {
-        return undefined;
+      if (moduleId instanceof NumberValue || moduleId instanceof StringValue) moduleIdValue = moduleId.value;
+      else return performCall();
+      // call require(...); this might cause calls to the define function
+      let res = this._callRequireAndRecord(moduleIdValue, performCall);
+      if (F.$Realm.eagerlyRequireModuleDependencies) {
+        // all dependencies of the required module should now be known
+        let dependencies = this.modules.moduleDependencies.get(moduleIdValue);
+        if (dependencies === undefined)
+          this.modules.logger.logError(moduleId, `Cannot resolve module dependencies for ${moduleIdValue.toString()}.`);
+        else
+          for (let dependency of dependencies)
+            if (dependency instanceof NumberValue || dependency instanceof StringValue)
+              this.modules.tryInitializeModule(dependency.value, `Eager initialization of module ${dependency.value}`);
```

Comment:
What happens when dependency is not a number or string? What happens if tryInitializeModule does not succeed? Since you throw away the return value of tryInitializeModule and it has no effect on the global state, it is not clear what the point of the call is. Is there a side-effect, such as an error message? Some comments about your intent here would be very helpful.